### ISSUE TITLE
Missing <graphics> for PDF figures under certain conditions

### ISSUE THREAD

**Issue by joseph-long** [brucemiller/LaTeXML#1979]
I'm trying to parse LaTeX from arXiv and determine the "most referenced" figure in a document. I've got `latexml --includestyles article.tex > article.xml` to produce an XML document, I can find `<ref>` and `<figure>` elements in the XML, and it seems like everything is working... except that the figure does not include a `<graphics>` element when the image in question is a PDF.

I pared down the LaTeX to get a minimal example. It looks like the (SPIE journal template recommended) nesting of environments inside `figure` may be responsible.

Inspecting the `<figure>` elements produced (below) note that the first figure produces a `<graphics>` element for its associated PDF, the second using SPIE's format does not, and the third using SPIE's format with a PNG does.

Is there a workaround?

**main.tex**

```latex
\documentclass{article}
\usepackage{graphicx}
\begin{document}
\begin{figure}
    \includegraphics[width=\textwidth]{dataflow.pdf}
    \label{fig:dataflow-test}
\end{figure}
\begin{figure}
    \begin{center}
    \begin{tabular}{c}
    \includegraphics[width=0.99\textwidth]{dataflow.pdf}
    \end{tabular}
    \end{center}
    \caption[dataflow]
    { \label{fig:dataflow} Command and data flow within MagAO-X}
\end{figure}
\begin{figure}
    \begin{center}
    \begin{tabular}{c}
    \includegraphics[width=\textwidth]{magao-x_web_ui_working.png}
    \end{tabular}
    \end{center}
    \caption[webui]
    { \label{fig:webui} The MagAO-X web interface}
\end{figure}
\end{document}
```

**main.xml**

```xml
<?xml version="1.0" encoding="UTF-8"?>
<?latexml searchpaths="/Users/josephlong/devel/chyme/mre"?>
<?latexml class="article"?>
<?latexml package="graphicx"?>
<?latexml RelaxNGSchema="LaTeXML"?>
<document xmlns="http://dlmf.nist.gov/LaTeXML" labels="LABEL:fig:dataflow-test" xml:id="id1">
  <resource src="LaTeXML.css" type="text/css"/>
  <resource src="ltx-article.css" type="text/css"/>
  <figure xml:id="fig1">
    <graphics candidates="dataflow.pdf" graphic="dataflow.pdf" options="width=433.62pt" xml:id="g1"/>
  </figure>
  <figure inlist="lof" labels="LABEL:fig:dataflow" xml:id="S0.F1">
    <tags>
      <tag>Figure 1</tag>
      <tag role="refnum">1</tag>
      <tag role="typerefnum">Figure 1</tag>
    </tags>
    <toccaption><tag close=" ">1</tag>dataflow</toccaption>
    <caption><tag close=": ">Figure 1</tag>  Command and data flow within MagAO-X</caption>
  </figure>
  <figure inlist="lof" labels="LABEL:fig:webui" xml:id="S0.F2">
    <tags>
      <tag>Figure 2</tag>
      <tag role="refnum">2</tag>
      <tag role="typerefnum">Figure 2</tag>
    </tags>
    <tabular class="ltx_centering" vattach="middle">
      <tbody>
        <tr>
          <td align="center"><graphics candidates="magao-x_web_ui_working.png" graphic="magao-x_web_ui_working.png" options="width=433.62pt" xml:id="S0.F2.g1"/></td>
        </tr>
      </tbody>
    </tabular>
<!--  %**** main.tex Line 25 **** -->    <toccaption><tag close=" ">2</tag>webui</toccaption>
    <caption><tag close=": ">Figure 2</tag>  The MagAO-X web interface</caption>
  </figure>
</document>
```

**Comment by joseph-long** [comment]
Also, possibly a separate issue report, but the first `<figure>` doesn't get a `labels=` attribute..

**Comment by dginev** [comment]
Indeed, I can reproduce the bug - thank you for reporting it @joseph-long

**Comment by joseph-long** [comment]
If you could give me a hint about where in the codebase to look, I could attempt a fix and make a pull request. I had a look myself, but it's kind of overwhelming.

**Comment by dginev** [comment]
@joseph-long I hear you. This is actually the kind of issue where I would first ask if anything here rings a bell to @brucemiller , who may already have a sharp instinct what went wrong in the latexml logic.

I think this may be a subtle case where the "empty table pruning" logic that was recently enhanced may end up pruning a `ltx:graphics` element, but that is just a guess on my part.

**Comment by joseph-long** [comment]
Good thinking; if it's a regression I might be able to find the change and *also* get a version that works for this project for me, already. I'll put that on my to-do list.

**Comment by joseph-long** [comment]
... relatedly, how the heck does one run LaTeXML in-place from a git checkout? I'm stuck on step 1 :(

**Comment by dginev** [comment]
> ... relatedly, how the heck does one run LaTeXML in-place from a git checkout? I'm stuck on step 1 :(

The old school perl way is:
```
$ perl Makefile.PL
$ make
$ perl -Iblib/lib blib/script/latexml ...
```

What I usually do in a hurry is to use [cpanminus](https://metacpan.org/dist/App-cpanminus/view/bin/cpanm), which allows a quick local install via `cpanm . --notest`.

You can see the example steps to get there, with all dependencies needed in Ubuntu, in this Dockerfile for instance: https://github.com/dginev/ar5ivist/blob/main/Dockerfile#L68-L72

**Comment by joseph-long** [comment]
Thanks for the tips.. I think I'll try making a container next. I got stuck on this on my mac:

```
Can't load '/opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/auto/XML/LibXML/LibXML.bundle' for module XML::LibXML: dlopen(/opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/auto/XML/LibXML/LibXML.bundle, 0x0001): Library not loaded: '@rpath/libicui18n.70.dylib'
  Referenced from: '/opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/auto/XML/LibXML/LibXML.bundle'
  Reason: tried: '/usr/local/lib/libicui18n.70.dylib' (no such file), '/usr/lib/libicui18n.70.dylib' (no such file) at /opt/homebrew/Cellar/perl/5.34.0/lib/perl5/5.34.0/XSLoader.pm line 96.
 at /opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/XML/LibXML.pm line 150.
```

After following the 'what provides libicui18n' -> 'how do I install an older version in homebrew' -> etc. rabbit hole for a bit, I gave up 🙃

**Comment by SWHL** [comment]
> Thanks for the tips.. I think I'll try making a container next. I got stuck on this on my mac:
> 
> ```
> Can't load '/opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/auto/XML/LibXML/LibXML.bundle' for module XML::LibXML: dlopen(/opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/auto/XML/LibXML/LibXML.bundle, 0x0001): Library not loaded: '@rpath/libicui18n.70.dylib'
>   Referenced from: '/opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/auto/XML/LibXML/LibXML.bundle'
>   Reason: tried: '/usr/local/lib/libicui18n.70.dylib' (no such file), '/usr/lib/libicui18n.70.dylib' (no such file) at /opt/homebrew/Cellar/perl/5.34.0/lib/perl5/5.34.0/XSLoader.pm line 96.
>  at /opt/homebrew/Cellar/perl/5.34.0/lib/perl5/site_perl/5.34.0/darwin-thread-multi-2level/XML/LibXML.pm line 150.
> ```
> 
> After following the 'what provides libicui18n' -> 'how do I install an older version in homebrew' -> etc. rabbit hole for a bit, I gave up 🙃

I met the same problem.

**Comment by dginev** [comment]
I see this as fixed in the latest master branch, commit 6ee12be9. I suspect #2286 or one of the related recent PRs resolved it.
